node/cn/backend.go

```diff
@@ -344,7 +345,19 @@ func New(ctx *node.ServiceContext, config *Config) (*CN, error) {
 	}
 	// TODO-Kaia-ServiceChain: add account creation prevention in the txPool if TxTypeAccountCreation is supported.
 	config.TxPool.NoAccountCreation = config.NoAccountCreation
-	cn.txPool = blockchain.NewTxPool(config.TxPool, cn.chainConfig, bc, mGov)
+
+	mGasless := gasless_impl.NewGaslessModule()
+	statedb, err := bc.State()
```

Comment:
You can't pass the statedb to gasless module and use it forever. `bc.State()` returns the latest state "at the point of node startup". And StateDB represents a frozen state at a certain block then never changes. In the module you must either
1. receive statedb argument to getter functions
2. call `g.Chain.State`every time to retrieve the latest state
3. Don't use statedb. Give up the checks that depend on nonce. Let TxPool deal with the nonce. (perhaps promoteTx is already checking tx.nonce == getNonce(tx.from)?)

Reply:
fyi, nonce check was to prevent multiple gasless txs from same sender in a block, thus allowing only one gasless tx pair at the getNonce(user).

Reply:
Oh, you're right @ian. Number 3 is not an option, because we specifically want to allow only one gasless tx per sender. promoteTx would allow multiple consecutive txs.

Between 1 and 2, I'd prefer 1 because you can simply pass `pool.currentState` as the argument.
- `chain.State()` is quite costly as it allocates multiple memory objects in `state.New`. Considering that the promotion check is performed for every transaction, the cost could be non-negligible.
- This makes the gasless module simpler because you can forget about states when testing/using any other getter functions.
- The same trick has been used by `engine.Finalize()`, `StateProcessor.Process()`.

Reply:
Or can we use reset hook as TxPoolModule on executing TxPool.reset? Statedb of gasless module can be updated every `Reset` execution. This function is also used for gov module.

https://github.com/kaiachain/kaia/blob/ab4a78d98caa22d2600fb16dbd05bcb94739e4a7/blockchain/tx_pool.go#L504-L507

Reply:
While you can call `TxPoolModule.Reset()` in `txpool.reset()`, you can also do the same in `ExecutionModule.PostInsertBlock()`. In that case, you may maintain the gasless module's internal variable `g.currentState`.

Note that reset and PostInsertBlock are called in very similar timings.
- insertChain -> PostInsertBlock
- insertChain -> PostChainEvents -> pool.chainHeadCh -> pool.reset

Reply:
I see. In this time, I select Reset of txpool because the main logic of gasless module is in the tx pool.